libs/utils/analysis/frequency_analysis.py

```diff
@@ -163,6 +163,127 @@ def _dfg_cpu_frequency_transition_rate(self, cpu):
+                if last_state == 0:
+                    freq.loc[index, 'frequency'] = 0
+                else:
+                    freq.loc[index, 'frequency'] = last_freq
+
+            if row.state == 1:
+                freq.loc[index, 'frequency'] = last_freq
+                last_state = row.state
+
+            if row.state == 0:
+                freq.loc[index, 'frequency'] = 0
+                last_state = row.state
+
+
+        fig = plt.figure(figsize=(16,8))
+        gs = gridspec.GridSpec(2, 1, height_ratios=[8, 1])
+        freq_axis = plt.subplot(gs[0])
+        state_axis = plt.subplot(gs[1])
+        plt.suptitle(title, y=.97, fontsize=16, horizontalalignment='center')
+
+        #plot frequency information
+        freq_axis.set_title("Clock frequency for " + clk)
+        freq_axis.set_ylim(0, max_freq * 1.1)
+        if len(freq) > 0:
+            freq['frequency'].plot(style=['b-'], ax=freq_axis, drawstyle='steps-post', alpha=0.4)
+        else:
+            self._log.warning('NO frequency events to plot')
+
+        freq_axis.set_xlim(self._trace.x_min, self._trace.x_max)
+        freq_axis.set_ylabel('MHz')
+        freq_axis.set_xlabel('')
+        freq_axis.grid(True)
+
+        #figure out when clocks are on and off
+        onoff = freq.loc[freq.state != -1, :]
+        first_onoff = self._trace.x_max
+        if len(onoff):
+            #edge detect when the state changes
+            onoff['statechange'] = onoff['state'].diff()
+            onoff = onoff[onoff.statechange != 0]
+
+            #compute delta as the time period between state changes
+            onoff['delta'] = (onoff['timestamp'] - onoff['timestamp'].shift()).fillna(0).shift(-1)
+            onoff.iloc[-1, onoff.columns.get_loc('delta')] = self._trace.x_max - onoff.iloc[-1].timestamp
+
+            #plot state on as green and off as red
+            enable_events = onoff[onoff.state == 1]
+            disable_events = onoff[onoff.state == 0]
+            state_axis.hlines([0] * len(enable_events),
+                          enable_events['timestamp'], enable_events['timestamp'] + enable_events['delta'],
+                          linewidth = 10.0, label='clock on', color='green')
+            state_axis.hlines([0] * len(disable_events),
+                          disable_events['timestamp'], disable_events['timestamp'] + disable_events['delta'],
+                          linewidth = 10.0, label='clock off', color='red')
+            first_onoff = onoff.iloc[0].timestamp
+
+        #plot time period that the clock state was unknown from the trace
+        state_axis.hlines(0, 0, first_onoff, linewidth = 1.0, label='indeterminate clock state', linestyle='--')
+        state_axis.set_yticks([])
+        state_axis.set_xlabel('seconds')
+        state_axis.set_xlim(self._trace.x_min, self._trace.x_max)
+        state_axis.legend(bbox_to_anchor=(0., 1.02, 1., 0.102), loc=3, ncol=3, mode='expand')
+
+        figname = '{}/{}{}.png'\
```

Comment:
`os.path.join()` would be better.

Reply:
done, the other functions could use an update too as a followup.